### DISCUSSION TITLE
[Ineligible for community support] How to configure LDAP tag_queries in new format??

### DISCUSSION THREAD

**Discussion by jsimao71** [rabbitmq/rabbitmq-server#10514] [Questions] [answered]
### Describe the bug

Configuration of LDAP tag_queries in new RabbitMQ configuration format, does not seam to work or it is not documented.
In traditional configuration this is the supported configuration:
[
  {rabbitmq_auth_backend_ldap, [
    {tag_queries, [{administrator, {constant, false}}]} 
  ]}
].

But what is the equivalent configuration in new format?!?
This example file does not have (it is commented with broken syntax):
Is there any other way to configure this when using bitnami rabbitmq helmchart?!?
Using:
+ new simplified configuration
+  set properties of the server with rabbitmqctl eval
+ env vars
+ other way
...

### Expected behavior

Should be possible to configure ldap tag_queries with new configuration format.
e.g.
auth_ldap.tag_queries = "[{administrator, {constant, false}}]"

### Additional context

We are using bitnami rabbitmq helmchart:
https://github.com/bitnami/charts/tree/main/bitnami/rabbitmq

**Comment by lukebakken** [comment]
You must use the `advanced.config` file to configure `tag_queries`, as documented here:

https://www.rabbitmq.com/ldap.html#authorisation

**Reply by jsimao71** [reply]
We have added this items to the bitnami helmchart to have the most permissive possible policy:
extraConfiguration: "auth_ldap.log = true"
advancedConfiguration: "[{rabbitmq_auth_backend_ldap, [{vhost_access_query, {constant, false}}, {resource_access_query, {constant, false}},{topic_access_query, {constant, false}},{tag_queries , [{administrator, {constant,true}}]}]}]."

**We setup the most possible permissive policy to check if it works, but still no luck!!**

This create a secret with the two entries:
_advanced.config: '[{rabbitmq_auth_backend_ldap, [{vhost_access_query, {constant, false}},
   {resource_access_query, {constant, false}},{topic_access_query, {constant, false}},{tag_queries
   , [{administrator, {constant,true}}]}]}].'
 rabbitmq.conf: |-
   ## Username and password
   ##
   default_user = rabbitmq
   default_pass = CHANGEME
   ## Clustering
   ##
   cluster_formation.peer_discovery_backend  = rabbit_peer_discovery_k8s
   cluster_formation.k8s.host = kubernetes.default
   cluster_formation.node_cleanup.interval = 10
   cluster_formation.node_cleanup.only_log_warning = true
   cluster_partition_handling = autoheal
   # queue master locator
   queue_master_locator = min-masters
   # enable guest user
   loopback_users.guest = false
   auth_ldap.log = true
   auth_backends.1.authn = ldap
   auth_backends.1.authz = internal
   auth_backends.2 = internal
   auth_ldap.servers.1 = 10.220.163.22
   auth_ldap.port = 389
   auth_ldap.dn_lookup_base = ou=Users,ou=DEV02,dc=oam,dc=dev02
   auth_ldap.dn_lookup_attribute = cn
   auth_ldap.dn_lookup_bind.user_dn = cn=test,ou=Users,ou=DEV02,dc=oam,dc=dev02
   auth_ldap.dn_lookup_bind.password = test123
   ## Prometheus metrics
   ##
   prometheus.tcp.port = 9419
   ## Memory Threshold
   ##
   total_memory_available_override_value = 4294967296
   vm_memory_high_watermark.absolute = 4800M_

Logs show that rabbitmq picked up the both config files:
   Config file(s): /opt/bitnami/rabbitmq/etc/rabbitmq/advanced.config
                   /opt/bitnami/rabbitmq/etc/rabbitmq/rabbitmq.conf

checking content of file:
cat /opt/bitnami/rabbitmq/etc/rabbitmq/advanced.config
[{rabbitmq_auth_backend_ldap, [{vhost_access_query, {constant, false}}, {resource_access_query, {constant, false}},{topic_access_query, {constant, false}},{tag_queries , [{administrator, {constant,true}}]}]}].


On login:
2024-02-14 10:52:54.275459+00:00 [info] <0.394.0>     LDAP bind succeeded: cn=xxxx,ou=xxxx,ou=xxxx,dc=xxxx,dc=xxxx
 2024-02-14 10:52:54.275664+00:00 [info] <0.394.0>         LDAP filling template "${username}" with
 2024-02-14 10:52:54.275664+00:00 [info] <0.394.0>             [{username,<<"test">>}]
 2024-02-14 10:52:54.275857+00:00 [info] <0.394.0>         LDAP template result: "test"
 2024-02-14 10:52:54.276884+00:00 [info] <0.394.0>     LDAP DN lookup: test -> cn=test,ou=Users,ou=DEV02,dc=oam,dc=dev02
 2024-02-14 10:52:54.277849+00:00 [info] <0.394.0>     LDAP bind succeeded: cn=xxxx,ou=xxxx,ou=xxxx,dc=xxxx,dc=xxxx
 2024-02-14 10:52:54.277926+00:00 [info] <0.394.0>     LDAP CHECK: does test have tag administrator?
 2024-02-14 10:52:54.277982+00:00 [info] <0.394.0>     LDAP evaluating query: {constant,true}
 2024-02-14 10:52:54.278020+00:00 [info] <0.394.0>     LDAP evaluated constant: true
 2024-02-14 10:52:54.278054+00:00 [info] <0.394.0>     LDAP DECISION: does test have tag administrator? true
 2024-02-14 10:52:54.278128+00:00 [info] <0.1682.0> LDAP DECISION: login for test: ok
 2024-02-14 10:52:54.278195+00:00 [warning] <0.1682.0> HTTP access denied: user 'test' - invalid credentials

So it seems that the tag administrator was assigned to user (test), but still was not allowed to login.
@lukebakken  what is missing in this advanced configuration?!?! Any idea?!?

Thanks in advance,
Jorge.

**Reply by jsimao71** [reply]
We also tried with this advanced configuration, same issue:
[{rabbitmq_auth_backend_ldap, [{vhost_access_query, {constant, true}}, {resource_access_query, {constant, true}},{topic_access_query, {constant, true}},{tag_queries , [{administrator, {constant,true}}]}]}].

(i.e. keep all setting with default, except the tag_queries to allow administrator tag)

Any suggestion @lukebakken  how to change the config file?!? Some missing bit?!?

2024-02-14 11:02:50.879280+00:00 [info] <0.777.0> LDAP CHECK: login for test
 2024-02-14 11:02:50.884172+00:00 [info] <0.394.0>     LDAP bind succeeded: cn=xxxx,ou=xxxx,ou=xxxx,dc=xxxx,dc=xxxx
 2024-02-14 11:02:50.884599+00:00 [info] <0.394.0>         LDAP filling template "${username}" with
 2024-02-14 11:02:50.884599+00:00 [info] <0.394.0>             [{username,<<"test">>}]
 2024-02-14 11:02:50.884828+00:00 [info] <0.394.0>         LDAP template result: "test"
 2024-02-14 11:02:50.886044+00:00 [info] <0.394.0>     LDAP DN lookup: test -> cn=test,ou=Users,ou=DEV02,dc=oam,dc=dev02
 2024-02-14 11:02:50.887139+00:00 [info] <0.394.0>     LDAP bind succeeded: cn=xxxx,ou=xxxx,ou=xxxx,dc=xxxx,dc=xxxx
 2024-02-14 11:02:50.887341+00:00 [info] <0.394.0>     LDAP CHECK: does test have tag administrator?
 2024-02-14 11:02:50.887482+00:00 [info] <0.394.0>     LDAP evaluating query: {constant,true}
 2024-02-14 11:02:50.887532+00:00 [info] <0.394.0>     LDAP evaluated constant: true
 2024-02-14 11:02:50.887581+00:00 [info] <0.394.0>     LDAP DECISION: does test have tag administrator? true
 2024-02-14 11:02:50.887638+00:00 [info] <0.777.0> LDAP DECISION: login for test: ok
 2024-02-14 11:02:50.887727+00:00 [warning] <0.777.0> HTTP access denied: user 'test' - invalid credentials

**Reply by lukebakken** [reply]
You need to _ATTACH_ your _COMPLETE_ `rabbitmq.conf` and `advanced.config` files. Copy and paste will not suffice.

You also haven't told us the version you're using of the following software:
* RabbitMQ
* Erlang
* Operating system and version

**Reply by jsimao71** [reply]
@lukebakken zip with the two files in attachment.
[rabbitmq-ldap.zip](https://github.com/rabbitmq/rabbitmq-server/files/14284337/rabbitmq-ldap.zip)
rabbitmq.conf and advanced.config

Details of deployment below:

<details>

```
 rabbitmqctl status
Status of node rabbit@rabbitmq2-0.rabbitmq2-headless.rabbitmq2.svc.cluster.local ...
Runtime

OS PID: 60
OS: Linux
Uptime (seconds): 23635
Is under maintenance?: false
RabbitMQ version: 3.10.8
RabbitMQ release series support status: supported
Node name: rabbit@rabbitmq2-0.rabbitmq2-headless.rabbitmq2.svc.cluster.local
Erlang configuration: Erlang/OTP 25 [erts-13.1.1] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]
Crypto library: OpenSSL 1.1.1n  15 Mar 2022
Erlang processes: 493 used, 1048576 limit
Scheduler run queue: 1
Cluster heartbeat timeout (net_ticktime): 60

Plugins

Enabled plugin file: /opt/bitnami/rabbitmq/etc/rabbitmq/enabled_plugins
Enabled plugins:

 * rabbitmq_prometheus
 * rabbitmq_auth_backend_ldap
 * rabbitmq_peer_discovery_k8s
 * accept
 * rabbitmq_peer_discovery_common
 * prometheus
 * rabbitmq_management
 * amqp_client
 * rabbitmq_web_dispatch
 * cowboy
 * cowlib
 * rabbitmq_management_agent

Data directory

Node data directory: /bitnami/rabbitmq/mnesia/rabbit@rabbitmq2-0.rabbitmq2-headless.rabbitmq2.svc.cluster.local
Raft data directory: /bitnami/rabbitmq/mnesia/rabbit@rabbitmq2-0.rabbitmq2-headless.rabbitmq2.svc.cluster.local/quorum/rabbit@rabbitmq2-0.rabbitmq2-headless.rabbitmq2.svc.cluster.local

Config files

 * /opt/bitnami/rabbitmq/etc/rabbitmq/advanced.config
 * /opt/bitnami/rabbitmq/etc/rabbitmq/rabbitmq.conf

Log file(s)

 * /opt/bitnami/rabbitmq/var/log/rabbitmq/rabbit@rabbitmq2-0.rabbitmq2-headless.rabbitmq2.svc.cluster.local_upgrade.log
 * <stdout>

Alarms

(none)

Memory

Total memory used: 0.1614 gb
Calculation strategy: rss
Memory high watermark setting: 5.0332 gb, computed to: 4.295 gb

reserved_unallocated: 0.0821 gb (50.9 %)
code: 0.0357 gb (22.15 %)
other_system: 0.0216 gb (13.37 %)
other_proc: 0.0193 gb (11.94 %)
other_ets: 0.002 gb (1.24 %)
metrics: 0.0017 gb (1.08 %)
atom: 0.0015 gb (0.95 %)
plugins: 0.0014 gb (0.89 %)
binary: 0.0003 gb (0.17 %)
mgmt_db: 0.0002 gb (0.14 %)
mnesia: 0.0001 gb (0.06 %)
connection_other: 0.0 gb (0.03 %)
msg_index: 0.0 gb (0.02 %)
quorum_ets: 0.0 gb (0.02 %)
quorum_queue_dlx_procs: 0.0 gb (0.0 %)
stream_queue_procs: 0.0 gb (0.0 %)
stream_queue_replica_reader_procs: 0.0 gb (0.0 %)
allocated_unused: 0.0 gb (0.0 %)
connection_channels: 0.0 gb (0.0 %)
connection_readers: 0.0 gb (0.0 %)
connection_writers: 0.0 gb (0.0 %)
queue_procs: 0.0 gb (0.0 %)
queue_slave_procs: 0.0 gb (0.0 %)
quorum_queue_procs: 0.0 gb (0.0 %)
stream_queue_coordinator_procs: 0.0 gb (0.0 %)

File Descriptors

Total: 2, limit: 1048479
Sockets: 0, limit: 943629

Free Disk Space

Low free disk space watermark: 0.05 gb
Free disk space: 7.9038 gb

Totals

Connection count: 0
Queue count: 0
Virtual host count: 1

Listeners

Interface: [::], port: 15672, protocol: http, purpose: HTTP API
Interface: [::], port: 9419, protocol: http/prometheus, purpose: Prometheus exporter API over HTTP
Interface: [::], port: 25672, protocol: clustering, purpose: inter-node and CLI tool communication
Interface: [::], port: 5672, protocol: amqp, purpose: AMQP 0-9-1 and AMQP 1.0
```
</details>

**Reply by lukebakken** [reply]
Please note how I used `<details> ... </details>` to collapse the wall of text you pasted. Large amounts of pasted text make it difficult to follow a discussion here, or anywhere online.

Why are you also using the internal auth backend this way? My guess is that your `test` user does not exist there.

```
auth_backends.1.authn = ldap
auth_backends.1.authz = internal
auth_backends.2 = internal
```

**Reply by jsimao71** [reply]
Sure!! If we create the user test with some dummy creds, then it works with the LDAP creds.
But we would like to be able to login users from LDAP (there are lots of them and are dynamic), without having to pre-create then and still allow for some internal users.
How to authenticate LDAP test users without having to pre-created it?!? Do have an idea @lukebakken ?!?

**Reply by jsimao71** [reply]
BTW, these `auth_backends.*` settings were done by bitnami chart did automatically.

**Reply by jsimao71** [reply]
@lukebakken  we found a way. Thanks for your help.
With this advanced config:
(2 times ldap: ldap for auth, and authz as first provider, and internal as second provider)
```
[{rabbit,[{auth_backends, [{rabbit_auth_backend_ldap,rabbit_auth_backend_ldap},rabbit_auth_backend_internal]}]},{rabbitmq_auth_backend_ldap, [{vhost_access_query, {constant, true}}, {resource_access_query, {constant, true}},{topic_access_query, {constant, true}},{tag_queries , [{administrator, {constant,true}}]}]}].
```

**Reply by michaelklishin** [reply]
Is that snippet really easy to read? Do you expect that our team will take the time to format it for you?

``` erl
[
  {rabbit,[
    {auth_backends, [
      {rabbit_auth_backend_ldap, rabbit_auth_backend_ldap}, rabbit_auth_backend_internal]
    }]},

  {rabbitmq_auth_backend_ldap, [
    {vhost_access_query, {constant, true}}, 
    {resource_access_query, {constant, true}},
    {topic_access_query, {constant, true}},
    {tag_queries , [{administrator, {constant,true}}]}
  ]}
].
```

**Reply by michaelklishin** [reply]
The `rabbit.auth_backends` part can be moved to `rabbitmq.conf` as @lukebakken demonstrates above.

The `rabbitmq_auth_backend_ldap` part must remain in `advanced.config` as it does not have a `rabbitmq.conf` equivalent (it uses complex data structures with arbitrary nesting depth).

**Reply by michaelklishin** [reply]
Also, what @lukebakken's example above has is NOT an equivalent of

```
{rabbit_auth_backend_ldap, rabbit_auth_backend_ldap}, rabbit_auth_backend_internal]
```

The `advanced.config` version is an equivalent of

``` ini
auth_backends.1.authn = ldap
auth_backends.1.authz = ldap
auth_backends.2 = internal
```

which can be simplified to just

``` ini
auth_backends.1 = ldap
auth_backends.2 = internal
```

Therefore, "use LDAP first for both authN and authZ, then try the internal backend for both authN and authZ".

**Comment by michaelklishin** [comment, marked as answer]
I'm afraid our team does not have the resources to troubleshoot LDAP servers, schemas, queries for non-paying users.

You can [enable LDAP query logging](https://rabbitmq.com/ldap.html#troubleshooting) and even logging of completely unfiltered LDAP (client) traffic. That alone makes most common problems easy to spot.

If you still need help with troubleshooting LDAP queries and schema, [your org can buy commercial support](https://forms.clickup.com/43297829/f/199b15-647834/SVPNW4U9G69FQCS65P).

**Reply by jsimao71** [reply]
No worries @michaelklishin ! We solved the issue. Thanks!
